blueman/gui/DeviceSelectorDialog.py

```diff
@@ -1,56 +1,192 @@
-            self.selection = selected_device["Adapter"], selected_device
+        row_adapter_label = builder.get_widget("row_adapter_name", Gtk.Label)
+        adapter_name = adapter_path_to_name(adapter_path)
+        assert adapter_name is not None
+        row_adapter_label.set_markup(f"<i>({adapter_name})</i>")
+
+        self.device_icon_name = device_icon
+        self.description = alias
+        self.warning = warning
+
+    @property
+    def device_icon_name(self) -> str:
+        return self._row_icon.get_icon_name()[0]
+
+    @device_icon_name.setter
+    def device_icon_name(self, icon_name: str) -> None:
+        self._row_icon.set_from_icon_name(icon_name, size=Gtk.IconSize.SMALL_TOOLBAR)
+
+    @property
+    def description(self) -> str:
+        return self._row_alias_label.get_label()
+
+    @description.setter
+    def description(self, text: str) -> None:
+        self._row_alias_label.set_label(text)
+
+    @property
+    def warning(self) -> bool:
+        return self._row_warning.get_visible()
+
+    @warning.setter
+    def warning(self, warning: bool) -> None:
+        self._row_warning.set_visible(warning)
+
+
+class DeviceSelector:
+    selection: Optional[Tuple[ObjectPath, Optional[Device]]]
+
+    def __init__(self, adapter_name: str = "any") -> None:
+        self._rows: dict[ObjectPath, DeviceRow] = {}
+        self._default_adapter_name = adapter_name
+        builder = Builder("sendto-device-dialog.ui")
+        self.dialog = builder.get_widget("select_device_dialog", Gtk.Dialog)
+        self._adapter_combo = builder.get_widget("adapter_combo", Gtk.ComboBoxText)
+
+        self._discover_spinner = builder.get_widget("discover_toggle_spinner", Gtk.Spinner)
+
+        self._listbox = builder.get_widget("device_listbox", Gtk.ListBox)
+        self._listbox.set_filter_func(self.__list_filter_func, None)
+        self._listbox.connect("row-selected", self.__on_row_selected)
+        self._listbox.connect("button-press-event", self.__on_button_press)
+
+        self._adapter_combo.connect("changed", lambda _: self._listbox.invalidate_filter())
+
+    def add_adapter(self, object_path: ObjectPath) -> None:
+        name = adapter_path_to_name(object_path)
+        if name is None:
+            raise ValueError("Invalid adapter")
+        pos = int(name[-1]) + 1
+        self._adapter_combo.insert(pos, name, name)
+        if name == self._default_adapter_name:
+            self._adapter_combo.set_active_id(name)
+
+    def remove_adapter(self, object_path: ObjectPath) -> None:
+        name = adapter_path_to_name(object_path)
+        if name is None:
+            raise ValueError("Invalid adapter")
+
+        for key in self._rows:
+            if self._rows[key].adapter_path == object_path:
+                row = self._rows.pop(key)
+                self._listbox.remove(row)
+
+        if name == self._adapter_combo.get_active_id():
+            self._adapter_combo.set_active_id(name)
```

Comment:
What's the intention here? :thinking:

Reply:
making sure the combobox has something valid selected. I changed it to select `any` if the active adapter is removed.